src/librustc_mir/interpret/place.rs

```diff
@@ -1029,7 +1029,9 @@ where
     ) -> InterpResult<'tcx> {
         match dest.layout.variants {
             layout::Variants::Single { index } => {
-                assert_eq!(index, variant_index);
+                if index != variant_index {
+                    throw_unsup_format!("tried to write a discriminate value out of range")
```

Comment:
When can this happen? And why is this not UB?

Reply:
This is the edge case exposed by the `dist-powerpc-linux` builder. The real code is in `crc32-fast` 

https://github.com/srijs/rust-crc32fast/blob/v1.1.2/src/lib.rs#L72-L77

where `State::Specialized` is an enum variant that holds a `specialized::State`. On x86, that is defined as a struct with various fields. On other platforms, `specialized::State` is an empty enum. So the `State` enum's layout gets "collapsed" (I don't know what the right terminology here is) to a single variant enum. When const-prop runs, this assert fails because we try to process the `SetDiscriminant(_n) = const 1usize` statement but according to `dest.layout`, there's only one variant and our index is invalid.

The test I added triggers this assert on all platforms before the change and passes after.

Reply:
But why do we generate MIR saying `SetDiscriminant(_n) = const 1usize` if `1` is not a valid discriminant? And why does an empty enum end up as a `Single` variant when it does not have exactly 1 variant (but rather has 0). Seems like a MIR building bug?

But either way it seems like this should be UB.

Reply:
Oh, I see what you're saying. It does seem weird that we're generating that MIR, or at least, we *still* have that MIR by the time const prop runs. You can see the invalid discriminant being set around `bb3` in the test ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=5367b2b6982f9f2af129c494aa3ff853)).

Perhaps `SimplifyBranches` should take into account switches on invalid enum discriminats?

Reply:
Nah, this shouldn't be cleaned up by optimizations.

But OTOH, this seems as good a way as any for MIR building to indicate that the code here is unreachable. So it's not entirely unreasonable to build this MIR. But Miri should treat this as UB. :D

Reply:
cc @oli-obk Do you have thoughts about this? Pruning the dead MIR out seems like it might be a win.

Reply:
I am a fan of the compiler design principle where you make codegen (like MIR lowering) as simple as you can get away with and then let optimizations clean up.

Reply:
> Nah, this shouldn't be cleaned up by optimizations.

Why not? I agree with @wesleywiser that known dead code should be eliminated by some optimization pass (preferably before inlining, so we don't inline dead code). Note that code may have become dead after optimizations.

Additionally it is my opinion that just like const prop, dead code elimination should emit warnings, instead of silently throwing away code.